File: python/lib/rpc.py
""" 
This file is part of the Nervatura Framework
http://nervatura.com
Copyright © 2011-2022, Csaba Kappel
License: LGPLv3
https://raw.githubusercontent.com/nervatura/nervatura/master/LICENSE 
"""

import os
import grpc
from nervatura import api_pb2 as pb, api_pb2_grpc as api
import json

def client():
    channel = grpc.insecure_channel("localhost:"+os.getenv("NT_GRPC_PORT"))
    return api.APIStub(channel)

def decodeValue(values):
    item = {}
    for key in dict(values).keys():
        if dict(values)[key].text == "null":
            item[key] = None
        elif str(dict(values)[key]).startswith("text:"):
            item[key] = dict(values)[key].text
        elif str(dict(values)[key]).startswith("number:"):
            item[key] = dict(values)[key].number
        else:
            item[key] = dict(values)[key].boolean
    return item

def encodeValue(value):
    if type(value) == bool:
        return pb.Value(bolean=value)
    if type(value) in(int, float):
        return pb.Value(number=float(value))
    if type(value) == list:
        return pb.Value(text=",".join(value))
    return pb.Value(text=str(value))

def mapValue(values):
    item = {}
    for key in dict(values).keys():
        item[key] = encodeValue(values[key])
    return item

def DatabaseCreate(apiKey, options):
    metadata = [("x-api-key", apiKey)]
    try:
        response = client().DatabaseCreate(pb.RequestDatabaseCreate(
            alias=options["database"], demo=options["demo"]
        ), metadata=metadata)
    except grpc.RpcError as err:
        return [None, err.details()]
    except:
        return [None, "Internal Error"]
    result = []
    for item in response.details.items:
        result.append(decodeValue(item.values))
    return [result, None]

def UserLogin(options):
    try:
        response = client().UserLogin(pb.RequestUserLogin(
            username=options["username"] if "username" in options else "", 
            password=options["password"] if "password" in options else "", 
            database=options["database"] if "database" in options else ""
        ))
    except grpc.RpcError as err:
        return [None, err.details()]
    except:
        return [None, "Internal Error"]
    result = dict(
        token=response.token,
        engine=response.engine,
        version=response.version
    )
    return [result, None]

def TokenLogin(token):
    metadata = [("authorization", "Bearer "+token)]
    try:
        response = client().TokenLogin(pb.RequestEmpty(), metadata=metadata)
    except grpc.RpcError as err:
        return [None, err.details()]
    except:
        return [None, "Internal Error"]
    result = dict(
        id=response.id,	username=response.username,	empnumber=response.empnumber,	
        usergroup=response.usergroup, scope=response.scope, department=response.department
    )
    return [result, None]

def TokenRefresh(token):
    metadata = [("authorization", "Bearer "+token)]
    try:
        response = client().TokenRefresh(pb.RequestEmpty(), metadata=metadata)
    except grpc.RpcError as err:
        return [None, err.details()]
    except:
        return [None, "Internal Error"]
    result = dict(
        token=response.value
    )
    return [result, None]

def UserPassword(token, options):
    metadata = [("authorization", "Bearer "+token)]
    try:
        client().UserPassword(pb.RequestUserPassword(
          password=options["password"] if "password" in options else "", 
          confirm=options["confirm"] if "confirm" in options else "", 
          username=options["username"] if "username" in options else "", 
          custnumber=options["custnumber"] if "custnumber" in options else ""
        ), metadata=metadata)
    except grpc.RpcError as err:
        return [None, err.details()]
    except:
        return [None, "Internal Error"]
    return [None, None]

def Report(token, options):
    metadata = [("authorization", "Bearer "+token)]
    try:
        response = client().Report(pb.RequestReport(
            reportkey=options["reportkey"],
            orientation=pb.ReportOrientation.values()[pb.ReportOrientation.keys().index(options["orientation"])],
            size=pb.ReportSize.values()[pb.ReportSize.keys().index(options["size"])],
            output= 3 if options["output"] == "print" or options["output"] == "auto" or options["output"] == "pdf" else pb.ReportOutput.values()[pb.ReportOutput.keys().index(options["output"])],
            type=pb.ReportType.values()[pb.ReportType.keys().index("report_"+options["nervatype"])],
            refnumber=options["refnumber"], template=options["template"],
            filters=mapValue(options["filters"])
        ), metadata=metadata)
    except grpc.RpcError as err:
        return [None, err.details()]
    except:
        return [None, "Internal Error"]
    result = json.loads(response.value)
    return [result, None]

def ReportList(token, options):
    metadata = [("authorization", "Bearer "+token)]
    try:
        response = client().ReportList(pb.RequestReportList(
            label=options["label"]
        ), metadata=metadata)
    except grpc.RpcError as err:
        return [None, err.details()]
    except:
        return [None, "Internal Error"]
    result = []
    for item in response.items:
        result.append(dict(
            reportkey=item.reportkey,
            repname=item.repname,
            description=item.description,
            label=item.label,
            reptype=item.reptype,
            filename=item.filename,
            installed=item.installed
        ))
    return [result, None]

def ReportInstall(token, options):
    metadata = [("authorization", "Bearer "+token)]
    try:
        response = client().ReportInstall(pb.RequestReportInstall(
            reportkey=options["reportkey"]
        ), metadata=metadata)
    except grpc.RpcError as err:
        return [None, err.details()]
    except:
        return [None, "Internal Error"]
    return [response.id, None]

def ReportDelete(token, options):
    metadata = [("authorization", "Bearer "+token)]
    try:
        client().ReportDelete(pb.RequestReportDelete(
          reportkey=options["reportkey"]
        ), metadata=metadata)
    except grpc.RpcError as err:
        return [None, err.details()]
    except:
        return [None, "Internal Error"]
    return [None, None]

def Function(token, options):
    metadata = [("authorization", "Bearer "+token)]
    try:
        response = client().Function(pb.RequestFunction(
            key=options["key"],
            values=mapValue(options["values"])
        ), metadata=metadata)
    except grpc.RpcError as err:
        return [None, err.details()]
    except:
        return [None, "Internal Error"]
    result = json.loads(response.value)
    return [result, None]

def View(token, data):
    rpcOptions = []
    for v in data:
        values = []
        for value in v["values"]:
            values.append(encodeValue(value))
        rpcOptions.append(dict(
           key=v["key"], text=v["text"], values=values
        ))
    metadata = [("authorization", "Bearer "+token)]
    try:
        response = client().View(pb.RequestView(
            options=rpcOptions
        ), metadata=metadata)
    except grpc.RpcError as err:
        return [None, err.details()]
    except:
        return [None, "Internal Error"]
    result = {}
    for vkey in response.values:
        items = []
        for item in response.values[vkey].items:
            items.append(decodeValue(item.values))
        result[vkey] = items
    return [result, None]

def Get(token, options):
    metadata = [("authorization", "Bearer "+token)]
    try:
        response = client().Get(pb.RequestGet(
            nervatype=pb.DataType.values()[pb.DataType.keys().index(options["nervatype"])],
            metadata=options["metadata"],
            filter=options["filter"],
            ids=options["ids"]
        ), metadata=metadata)
    except grpc.RpcError as err:
        return [None, err.details()]
    except:
        return [None, "Internal Error"]
    result = []
    for msg in response.values:
        ritem = {}
        item = msg.ListFields()[0][1]
        for fld in item.ListFields():
            if fld[0].name == "metadata":
                ritem["metadata"] = []
                for md in fld[1]:
                    ritem["metadata"].append(dict(
                        id=md.id,
                        fieldname=md.fieldname,
                        value=md.value,
                        notes=md.notes,
                        fieldtype=md.fieldtype
                    ))
            else:
                ritem[fld[0].name] = fld[1]
        result.append(ritem)
    return [result, None]

def Update(token, options):
    metadata = [("authorization", "Bearer "+token)]
    items = []
    for row in options["data"]:
        item = { "values":{} }
        for fkey in row:
            if fkey == "keys":
                item["keys"] = mapValue(row["keys"])
            else:
                item["values"][fkey]=encodeValue(row[fkey])
        items.append(item)
    try:
        response = client().Update(pb.RequestUpdate(
            nervatype=pb.DataType.values()[pb.DataType.keys().index(options["nervatype"])],
            items=items
        ), metadata=metadata)
    except grpc.RpcError as err:
        return [None, err.details()]
    except:
        return [None, "Internal Error"]
    result = []
    for id in response.values:
        result.append(id)
    return [result, None]

def Delete(token, options):
    metadata = [("authorization", "Bearer "+token)]
    try:
        client().Delete(pb.RequestDelete(
            nervatype=pb.DataType.values()[pb.DataType.keys().index(options["nervatype"])],
            id=options["id"], key=options["key"]
        ), metadata=metadata)
    except grpc.RpcError as err:
        return [None, err.details()]
    except:
        return [None, "Internal Error"]
    return [None, None]

File: php/lib/rpc.php
<?php
/*
This file is part of the Nervatura Framework
http://nervatura.com
Copyright © 2011-2023, Csaba Kappel
License: LGPLv3
https://raw.githubusercontent.com/nervatura/nervatura/master/LICENSE
*/

class Rpc {
  function client(){
    return new Nervatura\APIClient('localhost:'.getenv('NT_GRPC_PORT'), [
      'credentials' => Grpc\ChannelCredentials::createInsecure(),
    ]);
  }

  function decodeValue($values) {
    $item = array();
    foreach ($values as $key => $value) {
      if($value->getValue() == 'boolean'){
        $item = array_merge($item, array($key => $value->getBoolean()));
      } else if($value->getValue() == 'number'){
        $item = array_merge($item, array($key => $value->getNumber()));
      } else if($value->getText() == 'null') {
        $item = array_merge($item, array($key => NULL));
      } else {
        $item = array_merge($item, array($key => $value->getText()));
      }
    }
    return $item;
  }

  function encodeValue($value) {
    $rpcValue = new Nervatura\Value();
    switch (gettype($value)) {
      case 'boolean':
        $rpcValue->setBoolean($value);
        break;

      case 'integer':
      case 'double':
        $rpcValue->setNumber($value);
        break;
      
      case 'array':
        $rpcValue->setText(implode(',', $value));
        break;
      
      default:
        $rpcValue->setText(strval($value));
        break;
    }
    return $rpcValue;
  }

  function mapValue($values) {
    $item = array();
    foreach ($values as $key => $value) {
      $item = array_merge($item, array($key => $this->encodeValue($value)));
    }
    return $item;
  }

  function checkValue($values) {
    foreach ($values as $key => $value) {
      switch (gettype($value)) {
        case 'boolean':
        case 'integer':
        case 'double':
        case 'string':
          break;

        case 'array':
          if(is_array($value) && !is_numeric(implode("", array_keys($value)))){
            return false;
          };
          break;
        
        default:
          return false;
      }
    }
    return true;
  }

  public function DatabaseCreate($api_key, $options){
    $metadata = ['x-api-key' => [$api_key]];
    $options['demo'] = ($options['demo'] == 'true');
    $request = new Nervatura\RequestDatabaseCreate($options);
    list($response, $status) = $this->client()->DatabaseCreate($request, $metadata)->wait();
    if ($status->code !== Grpc\STATUS_OK) {
      return array(
        'result' => NULL,
        'error' => $status->details
      );
    }
    $result = array();
    $items = $response->getDetails()->getItems();
    foreach ($items as $key => $value) {
      array_push($result, $this->decodeValue($value->getValues()));
    }
    return array(
      'result' => $result,
      'error' => NULL
    );
  }

  public function UserLogin($options){
    $request = new Nervatura\RequestUserLogin($options);
    list($response, $status) = $this->client()->UserLogin($request)->wait();
    if ($status->code !== Grpc\STATUS_OK) {
      return array(
        'result' => NULL,
        'error' => $status->details
      );
    }
    return array(
      'result' => array(
        'engine' => $response->getEngine(),
        'version' => $response->getVersion(),
        'token' => $response->getToken()
      ),
      'error' => NULL
    );
  }

  public function TokenLogin($token){
    $metadata = ['authorization' => ['Bearer '.$token]];
    $request = new Nervatura\RequestEmpty();
    list($response, $status) = $this->client()->TokenLogin($request, $metadata)->wait();
    if ($status->code !== Grpc\STATUS_OK) {
      return array(
        'result' => NULL,
        'error' => $status->details
      );
    }
    return array(
      'result' => array(
        'id' => $response->getId(),
        'username' => $response->getUsername(),
        'empnumber' => $response->getEmpnumber(),
        'usergroup' => $response->getUsergroup(),
        'scope' => $response->getScope(),
        'department' => $response->getDepartment()
      ),
      'error' => NULL
    );
  }

  public function TokenRefresh($token){
    $metadata = ['authorization' => ['Bearer '.$token]];
    $request = new Nervatura\RequestEmpty();
    list($response, $status) = $this->client()->TokenRefresh($request, $metadata)->wait();
    if ($status->code !== Grpc\STATUS_OK) {
      return array(
        'result' => NULL,
        'error' => $status->details
      );
    }
    return array(
      'result' => array(
        'token' => $response->getValue()
      ),
      'error' => NULL
    );
  }

  public function UserPassword($token, $options){
    $metadata = ['authorization' => ['Bearer '.$token]];
    $request = new Nervatura\RequestUserPassword($options);
    list($response, $status) = $this->client()->UserPassword($request, $metadata)->wait();
    if ($status->code !== Grpc\STATUS_OK) {
      return array(
        'result' => NULL,
        'error' => $status->details
      );
    }
    return array(
      'result' => NULL,
      'error' => NULL
    );
  }

  public function Report($token, $options){
    function output($output_name) {
      switch ($output_name) {
        case 'print':
        case 'auto':
        case 'pdf':
          return 3;
        default:
          return (new Nervatura\ReportOutput())->value($output_name);
      }
    }
    $metadata = ['authorization' => ['Bearer '.$token]];
    $request = new Nervatura\RequestReport(array(
      'reportkey' => $options['reportkey'],
      'orientation' => (new Nervatura\ReportOrientation())->value($options['orientation']),
      'size' => (new Nervatura\ReportSize())->value($options['size']),
      'output' => output($options['output']),
      'type' => (new Nervatura\ReportType())->value('report_'.$options['nervatype']),
      'refnumber' => $options['refnumber'],
      'template' => $options['template'],
      'filters' => $this->mapValue($options['filters'])
    ));
    list($response, $status) = $this->client()->Report($request, $metadata)->wait();
    if ($status->code !== Grpc\STATUS_OK) {
      return array(
        'result' => NULL,
        'error' => $status->details
      );
    }
    return array(
      'result' => json_decode($response->getValue(), true),
      'error' => NULL
    );
  }

  public function ReportList($token, $options){
    $metadata = ['authorization' => ['Bearer '.$token]];
    $request = new Nervatura\RequestReportList($options);
    list($response, $status) = $this->client()->ReportList($request, $metadata)->wait();
    if ($status->code !== Grpc\STATUS_OK) {
      return array(
        'result' => NULL,
        'error' => $status->details
      );
    }
    $result = array();
    $items = $response->getItems();
    foreach ($items as $key => $value) {
      array_push($result, array(
        'reportkey' => $value->getReportkey(),
        'repname' => $value->getRepname(),
        'description' => $value->getDescription(),
        'label' => $value->getLabel(),
        'reptype' => $value->getReptype(),
        'filename' => $value->getFilename(),
        'installed' => $value->getInstalled()
      ));
    }
    return array(
      'result' => $result,
      'error' => NULL
    );
  }

  public function ReportInstall($token, $options){
    $metadata = ['authorization' => ['Bearer '.$token]];
    $request = new Nervatura\RequestReportInstall($options);
    list($response, $status) = $this->client()->ReportInstall($request, $metadata)->wait();
    if ($status->code !== Grpc\STATUS_OK) {
      return array(
        'result' => NULL,
        'error' => $status->details
      );
    }
    return array(
      'result' => array(
        'id' => $response->getId()
      ),
      'error' => NULL
    );
  }

  public function ReportDelete($token, $options){
    $metadata = ['authorization' => ['Bearer '.$token]];
    $request = new Nervatura\RequestReportDelete($options);
    list($response, $status) = $this->client()->ReportDelete($request, $metadata)->wait();
    if ($status->code !== Grpc\STATUS_OK) {
      return array(
        'result' => NULL,
        'error' => $status->details
      );
    }
    return array(
      'result' => NULL,
      'error' => NULL
    );
  }

  public function Function($token, $options){
    $metadata = ['authorization' => ['Bearer '.$token]];
    $value = ''; $values = array();
    if(array_key_exists('values', $options)){
      if ($this->checkValue($options['values'])) {
        $values = $this->mapValue($options['values']);
      } else {
        $value = json_encode($options['values']);
      }
    }
    $request = new Nervatura\RequestFunction(array(
      'key' => $options['key'],
      'value' => $value,
      'values' => $values,
    ));
    list($response, $status) = $this->client()->Function($request, $metadata)->wait();
    if ($status->code !== Grpc\STATUS_OK) {
      return array(
        'result' => NULL,
        'error' => $status->details
      );
    }
    return array(
      'result' => json_decode($response->getValue(), true),
      'error' => NULL
    );
  }

  public function View($token, $data){
    $metadata = ['authorization' => ['Bearer '.$token]];
    $rpcOptions = array();
    foreach ($data as $ikey => $item) {
      $values = array();
      foreach ($item['values'] as $key => $value) {
        array_push($values, $this->encodeValue($value));
      }
      array_push($rpcOptions, new Nervatura\RequestView\Query(array(
        'key' => $item['key'],
        'text' => $item['text'],
        'values' => $values
      )));
    }
    $request = new Nervatura\RequestView(array(
      'options' => $rpcOptions
    ));
    list($response, $status) = $this->client()->View($request, $metadata)->wait();
    if ($status->code !== Grpc\STATUS_OK) {
      return array(
        'result' => NULL,
        'error' => $status->details
      );
    }
    $result = array();
    foreach ($response->getValues() as $ikey => $item) {
      $items = array();
      foreach ($item->getItems() as $vkey => $value) {
        array_push($items, $this->decodeValue($value->getValues()));
      }
      $result = array_merge($result, array($ikey => $items));
    }
    return array(
      'result' => $result,
      'error' => NULL
    );
  }

  public function Get($token, $options){
    $metadata = ['authorization' => ['Bearer '.$token]];
    $rpcOptions = array(
      'nervatype' => (new Nervatura\DataType())->value($options['nervatype'])
    );
    if(array_key_exists('metadata', $options)){
      $rpcOptions = array_merge($rpcOptions, array('metadata' => $options['metadata']));
    }
    if(array_key_exists('ids', $options)){
      if(!is_array($options['ids'])) {
        $rpcOptions = array_merge($rpcOptions, 
          array('ids' => json_decode('['.$options['ids'].']'))
        );
      } else {
        $rpcOptions = array_merge($rpcOptions, array('ids' => $options['ids']));
      }
    }
    if(array_key_exists('filter', $options)){
      $rpcOptions = array_merge($rpcOptions, array('filter' => explode('|', $options['filter'])));
    }
    $request = new Nervatura\RequestGet($rpcOptions);
    list($response, $status) = $this->client()->Get($request, $metadata)->wait();
    if ($status->code !== Grpc\STATUS_OK) {
      return array(
        'result' => NULL,
        'error' => $status->details
      );
    }
    $result = array();
    foreach ($response->getValues() as $mkey => $item) {
      array_push($result, json_decode($item->serializeToJsonString(),true)[$item->getValue()]);
    }
    return array(
      'result' => $result,
      'error' => NULL
    );
  }

  public function Update($token, $options){
    $metadata = ['authorization' => ['Bearer '.$token]];
    $items = array();
    foreach ($options['data'] as $rkey => $row) {
      $item = new Nervatura\RequestUpdate\Item();
      $values = array();
      foreach ($row as $key => $value) {
        if($key == 'keys') {
          $item->setKeys($this->mapValue($value));
        } else {
          $values = array_merge($values, array($key => $this->encodeValue($value)));
        }
      }
      $item->setValues($values);
      array_push($items, $item);
    }
    $request = new Nervatura\RequestUpdate(array(
      'nervatype' => (new Nervatura\DataType())->value($options['nervatype']),
      'items' => $items
    ));
    list($response, $status) = $this->client()->Update($request, $metadata)->wait();
    if ($status->code !== Grpc\STATUS_OK) {
      return array(
        'result' => NULL,
        'error' => $status->details
      );
    }
    $result = array();
    foreach ($response->getValues() as $key => $id) {
      array_push($result, $id);
    }
    return array(
      'result' => $result,
      'error' => NULL
    );
  }

  public function Delete($token, $options){
    $metadata = ['authorization' => ['Bearer '.$token]];
    $rpcOptions = array(
      'nervatype' => (new Nervatura\DataType())->value($options['nervatype'])
    );
    if(array_key_exists('id', $options)){
      $rpcOptions = array_merge($rpcOptions, array('id' => $options['id']));
    }
    if(array_key_exists('key', $options)){
      $rpcOptions = array_merge($rpcOptions, array('key' => $options['key']));
    }
    $request = new Nervatura\RequestDelete($rpcOptions);
    list($response, $status) = $this->client()->Delete($request, $metadata)->wait();
    if ($status->code !== Grpc\STATUS_OK) {
      return array(
        'result' => NULL,
        'error' => $status->details
      );
    }
    return array(
      'result' => NULL,
      'error' => NULL
    );
  }

}

?>